File: LeetCode/437.Path_Sum_III/437.Path_Sum_III.py
#!/usr/bin/env python3


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        res = 0

        def dfs(
            node, prefix_dic, prefixsum
        ):  # dictionary key:prefixsum: cnt, prefixsum
            nonlocal res
            # base case:
            if not node:
                return

            new_prefixsum = prefixsum + node.val
            if new_prefixsum - targetSum in prefix_dic:
                res += prefix_dic[new_prefixsum - targetSum]
            prefix_dic[new_prefixsum] += 1

            # go further subtree
            dfs(node.left, prefix_dic, new_prefixsum)
            dfs(node.right, prefix_dic, new_prefixsum)

            prefix_dic[new_prefixsum] -= 1

        dic = defaultdict(int)
        dic[0] = 1
        dfs(root, dic, 0)

        return res

File: LeetCode/437.Path_Sum_III/437.Path_Sum_III.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    private int count = 0;
    int target = 0;
    // store prefixSum and its count
    HashMap<Long, Integer> prefixSumMap = new HashMap<>();

    public int pathSum(TreeNode root, int targetSum) {
        // prefix sum + map
        target = targetSum;
        dfs(root, 0L);
        return count;
    }

    public void dfs(TreeNode node, long sum) {
        if (node == null) {
            return;
        }

        sum += node.val;

        if (sum == target) {
            count += 1;
        }

        count += prefixSumMap.getOrDefault(sum - target, 0);

        prefixSumMap.put(sum, prefixSumMap.getOrDefault(sum, 0) + 1);

        dfs(node.left, sum);
        dfs(node.right, sum);

        prefixSumMap.put(sum, prefixSumMap.get(sum) - 1);
    }
}